File: src/main/java/com/samsolutions/recipes/service/RecipeService.java
package com.samsolutions.recipes.service;

import com.samsolutions.recipes.dto.RecipeDTO;
import com.samsolutions.recipes.dto.UserDTO;
import com.samsolutions.recipes.dto.createRecipe.CreateRecipeDTO;
import com.samsolutions.recipes.dto.findByData.RecipeDataDTO;
import com.samsolutions.recipes.dto.findByIngredients.IngredientNameListDTO;
import com.samsolutions.recipes.model.RecipeEntity;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;
import java.util.UUID;

/**
 * @author kaminskiy.alexey
 * @since 2019.12
 */
public interface RecipeService {

    List<RecipeDTO> findAll();

    RecipeDTO create(RecipeDTO recipeDTO);

    RecipeDTO update(UUID uuid, RecipeDTO recipeDTO);

    RecipeDTO findByRecipeId(UUID uuid);

    CreateRecipeDTO updateRecipe(UUID uuid, CreateRecipeDTO recipeDTO);

    void positiveVote(UUID uuid);

    void negativeVote(UUID uuid);

    void removeById(UUID uuid);

    List<RecipeDTO> getByCategoryName(String categoryName, int page, int size, String sort, UUID... userId);

    int getCountAllRecipesInCategory(String categoryName);

    int getCountAllOwnRecipes(UUID authorId);

    CreateRecipeDTO createRecipeDTO(CreateRecipeDTO createRecipeDTO);

    CreateRecipeDTO getByRecipeId(UUID uuid);

    List<CreateRecipeDTO> getByAuthorId(UUID authorId, int page, int size, String sort);

    List<RecipeDTO> getByAuthorName(String name, int page, int size, String sort, UUID... userId);

    int getCountAllRecipesByAuthorName(String name);

    List<RecipeDTO> findAllByName(String name, int page, int size, String sort, UUID... userId);

    int getCountAllRecipesByName(String name);

    RecipeDTO getByNameAuthorId(String name, UUID uuid);

    List<RecipeDTO> findAllByIngredient(IngredientNameListDTO ingredientNameListDTO, int page, int size, String sort, UUID... userId);

    int getCountAllRecipesByIngredient();

    List<RecipeDTO> findAllByData(RecipeDataDTO recipeDataDTO, int page, int size, String sort, UUID... userId);

    int getCountAllRecipesByData();

    RecipeDTO savePhoto(UUID id, MultipartFile file) throws IOException;

    UserDTO getAuthorName(String authorId);

    //for createRecipe
    void saveCategoryRecipeEntityList(CreateRecipeDTO createRecipeDTO);

    List<UUID> saveCookingStepsEntityList(CreateRecipeDTO createRecipeDTO);

    void saveRecipeIngredientList(CreateRecipeDTO createRecipeDTO);

    //for UpdateRecipe
    void updateCategory(CreateRecipeDTO createRecipeDTO);

    List<UUID> updateCookingStepsEntityList(CreateRecipeDTO createRecipeDTO);

    void updateRecipeIngredientList(CreateRecipeDTO createRecipeDTO);

    //for getByRecipeId and getByAuthorId
    CreateRecipeDTO mapCategoryRecipeEntityListToDTO(UUID uuid, CreateRecipeDTO createRecipeDTO);

    CreateRecipeDTO mapCookingStepRecipeEntityListToDTO(RecipeEntity recipeEntity, CreateRecipeDTO createRecipeDTO);

    CreateRecipeDTO mapRecipeIngredientEntityListToDTO(RecipeEntity recipeEntity, CreateRecipeDTO createRecipeDTO);

}

File: frontend/src/app/service/recipe.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {Recipe} from "../model/recipe";
import {CreateRecipeDTO} from "../model/createRecipe/create-recipe-dto";
import {IngredientNameListDTO} from "../model/findByIngredients/ingredient-name-list-dto";
import {RecipeDataDTO} from "../model/findByData/recipe-data-dto";

@Injectable({
  providedIn: 'root'
})
export class RecipeService {

  private baseUrl: string;

  constructor(private http: HttpClient) {
    this.baseUrl = '/api/recipe';
  }

  public getRecipesByCategoryName(categoryName: string, userId: string, page: number, size: number, sort: string): Observable<Recipe[]> {
    return this.http.get<Recipe[]>(`${this.baseUrl}/categoryName/${categoryName}?page=${page}&size=${size}&sort=${sort}&userId=${userId}`);
  }

  public getRecipesByCategoryName2(categoryName: string, page: number, size: number, sort: string): Observable<Recipe[]> {
    return this.http.get<Recipe[]>(`${this.baseUrl}/categoryName/${categoryName}?page=${page}&size=${size}&sort=${sort}&userId=`);
  }

  public getCountAllRecipesInCategory(categoryName: string): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/countAllRecipesInCategory/${categoryName}`);
  }

  public getCountAllRecipesByIngredient(): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/countAllRecipesByIngredient`);
  }

  public getCountAllRecipesByData(): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/countAllRecipesByData`);
  }

  public getCountAllRecipesByName(categoryName: string): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/countAllRecipesByName/${categoryName}`);
  }

  public getCountAllRecipesByAuthorName(authorName: string): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/countAllRecipesByAuthorName/${authorName}`);
  }

  public getCountAllOwnRecipes(authorId: string): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/countAllOwnRecipes/${authorId}`);
  }

  public createRecipe(recipe: CreateRecipeDTO) {
    return this.http.post<CreateRecipeDTO>(`${this.baseUrl}/createRecipe`, recipe);
  }

  public getById(id: string): Observable<CreateRecipeDTO> {
    return this.http.get<CreateRecipeDTO>(`${this.baseUrl}/id/${id}`);
  }

  public getByAuthorId(id: string, page: number, size: number, sort: string): Observable<CreateRecipeDTO[]> {
    return this.http.get<CreateRecipeDTO[]>(`${this.baseUrl}/authorId/${id}?page=${page}&size=${size}&sort=${sort}`);
  }

  public delete(id: string): Observable<any> {
    return this.http.delete(`${this.baseUrl}/delete/${id}`, {responseType: 'text'});
  }

  public update(id: string, recipeData: any): Observable<CreateRecipeDTO> {
    return this.http.put<CreateRecipeDTO>(`${this.baseUrl}/updateRecipe/${id}`, recipeData);
  }

  public getByName(name: string, page: number, size: number, sort: string, userId: string): Observable<Recipe[]> {
    return this.http.get<Recipe[]>(`${this.baseUrl}/name/${name}?page=${page}&size=${size}&sort=${sort}&userId=${userId}`);
  }

  public getByName2(name: string, page: number, size: number, sort: string): Observable<Recipe[]> {
    return this.http.get<Recipe[]>(`${this.baseUrl}/name/${name}?page=${page}&size=${size}&sort=${sort}&userId=`);
  }

  public getByNameAndAuthor(name: string, id: string): Observable<Recipe> {
    return this.http.get<Recipe>(`${this.baseUrl}/nameAndAuthor/${name}/${id}`);
  }

  public getByAuthorName(name: string, page: number, size: number, sort: string, userId: string): Observable<Recipe[]> {
    return this.http.get<Recipe[]>(`${this.baseUrl}/authorName/${name}?page=${page}&size=${size}&sort=${sort}&userId=${userId}`);
  }

  public getByAuthorName2(name: string, page: number, size: number, sort: string): Observable<Recipe[]> {
    return this.http.get<Recipe[]>(`${this.baseUrl}/authorName/${name}?page=${page}&size=${size}&sort=${sort}&userId=`);
  }

  public findAllByIngredients(ingredientsList: IngredientNameListDTO, page: number, size: number, sort: string, userId: string) {
    return this.http.post<Recipe[]>(`${this.baseUrl}/ingredientName?page=${page}&size=${size}&sort=${sort}&userId=${userId}`, ingredientsList);
  }

  public findAllByIngredients2(ingredientsList: IngredientNameListDTO, page: number, size: number, sort: string) {
    return this.http.post<Recipe[]>(`${this.baseUrl}/ingredientName?page=${page}&size=${size}&sort=${sort}&userId=`, ingredientsList);
  }

  public findAllByData(recipeData: RecipeDataDTO, page: number, size: number, sort: string, userId: string) {
    return this.http.post<Recipe[]>(`${this.baseUrl}/data?page=${page}&size=${size}&sort=${sort}&userId=${userId}`, recipeData);
  }

  public findAllByData2(recipeData: RecipeDataDTO, page: number, size: number, sort: string) {
    return this.http.post<Recipe[]>(`${this.baseUrl}/data?page=${page}&size=${size}&sort=${sort}&userId=`, recipeData);
  }

  public addPhoto4Step(stepId: string, file: File) {
    const data = new FormData();
    data.append('file', file);
    console.log(data);
    fetch(`${this.baseUrl}/addPhoto4Step/${stepId}`, {
      method: 'POST',
      body: data
    }).then(r => {

    }).catch(r => {

    });
  }

  public addPhoto4Recipe(recipeId: string, file: File) {
    const data = new FormData();
    data.append('file', file);
    fetch(`${this.baseUrl}/addPhoto4Recipe/${recipeId}`, {
      method: 'POST',
      body: data
    }).then(r => {

    });
  }
}
